### PR TITLE
treewide: use stdenvNoCC when not building

### PR DESCRIPTION
**Pull request by siraben**
###### Motivation for this change
In many packages, the C compiler is not used, so stdenv could be stdenvNoCC.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by roberth** [comment]
Is it just me or is `stdenvNoCC` an ugly name? It contradicts itself. The _standard_ environment has a C compiler, so `stdenvNoCC` is _not_ standard. If we're going to use this more, what about `smallenv`?
Sorry if off-topic. Vote for `smallenv` with :+1: or :-1: reactions.

**Comment by siraben** [comment]
Naming discussion would be good in a separate issue, but I don't know if smallenv is any more descriptive than stdenvNoCC.

Would this PR introduce unexpected breakage? I haven't been able to test it because of the number of rebuilds.

**Comment by roberth** [comment]
I'd expect breakage, if any, to be in the form of failed builds, which isn't the end of the world.
It looks like you went for drvs with `dontBuild = true`. That seems like it would correlate well with not using a C compiler.
A random sample should be enough to decide whether this is good enough.

File: pkgs/applications/version-management/git-and-tools/diff-so-fancy/default.nix [hunk 1]
```diff
@@ -1,6 +1,6 @@
-{lib, stdenv, git, perl, ncurses, coreutils, fetchFromGitHub, makeWrapper, ...}:
+{lib, stdenvNoCC, git, perl, ncurses, coreutils, fetchFromGitHub, makeWrapper, ...}:
```
Comment (SuperSandro2000) [inline pkgs/applications/version-management/git-and-tools/diff-so-fancy/default.nix:1]:
```suggestion
{lib, stdenvNoCC, git, perl, ncurses, coreutils, fetchFromGitHub, makeWrapper }:
```

**Comment by SuperSandro2000** [comment]
```
nix-build -A blas
these derivations will be built:
  /nix/store/7s7q395a8zlga6ln7b57yrgy6bs32ksd-blas-3.drv
error: error: builder for '/nix/store/7s7q395a8zlga6ln7b57yrgy6bs32ksd-blas-3.drv' failed with exit code 127;
       last 4 log lines:
       > unpacking sources
       > patching sources
       > installing
       > /nix/store/257gr6liaypg83n9cn0lq53vxh6hb27c-stdenv-linux/setup: line 1318: -an: command not found
       For full logs, run 'nix log /nix/store/7s7q395a8zlga6ln7b57yrgy6bs32ksd-blas-3.drv'
```

**Comment by SuperSandro2000** [comment]
This is a semi-automatic executed nixpkgs-review with [nixpkgs-review-checks extension](https://github.com/SuperSandro2000/nixpkgs-review-checks). It is checked by a human on a best effort basis and does not build all packages (e.g. lumo, tensorflow or pytorch).
If you have any questions or problems please reach out to SuperSandro2000 on IRC.

Result of `nixpkgs-review pr 117451` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>

  <summary>23 packages failed to build and are new build failures:</summary>
  <ul>
<li>blas: log was empty</li>
<li>drawio: log was empty</li>
<li>equilux-theme: log was empty</li>
<li>factorio: log was empty</li>
<li>freetube: log was empty</li>
<li>i3-layout-manager: log was empty</li>
<li>jicofo: log was empty</li>
<li>jitsi-videobridge: log was empty</li>
<li>macse: log was empty</li>
<li>marwaita: log was empty</li>
<li>marwaita-manjaro: log was empty</li>
<li>marwaita-peppermint: log was empty</li>
<li>marwaita-pop_os: log was empty</li>
<li>marwaita-ubuntu: log was empty</li>
<li>materia-theme: log was empty</li>
<li>orchis: log was empty</li>
<li>skeu: log was empty</li>
<li>split2flac: log was empty</li>
<li>steamcmd: log was empty</li>
<li>streamlink-twitch-gui-bin: log was empty</li>
<li>sway-contrib.grimshot: log was empty</li>
<li>venta: log was empty</li>
<li>wmic-bin: log was empty</li>
  </ul>
</details>
<details>
<summary>
51 packages built:
</summary>
<ul>
<li>albatross</li>
<li>clearlooks-phenix</li>
<li>diff-so-fancy</li>
<li>docbook5_xsl</li>
<li>dxx-rebirth</li>
<li>exodus</li>
<li>fish</li>
<li>fzf-zsh</li>
<li>git-extras</li>
<li>git-filter-repo</li>
<li>git-my</li>
<li>git-radar</li>
<li>git-reparent</li>
<li>git-secrets</li>
<li>git-standup</li>
<li>git-sync</li>
<li>git-test</li>
<li>gitlab</li>
<li>gitolite</li>
<li>gnomeExtensions.appindicator</li>
<li>gnomeExtensions.material-shell</li>
<li>gnomeExtensions.noannoyance</li>
<li>gnomeExtensions.paperwm</li>
<li>gnomeExtensions.remove-dropdown-arrows</li>
<li>gnomeExtensions.sound-output-device-chooser</li>
<li>gradle-completion</li>
<li>grocy</li>
<li>jetty</li>
<li>jitsi-meet</li>
<li>lapack</li>
<li>longview</li>
<li>mpv</li>
<li>mustache-spec</li>
<li>numix-sx-gtk-theme</li>
<li>profont</li>
<li>pulseaudio-hsphfpd</li>
<li>real_time_config_quick_scan</li>
<li>redoflacs</li>
<li>roundcube</li>
<li>scheme-manpages</li>
<li>scummvm</li>
<li>sickbeard</li>
<li>sickgear</li>
<li>svn2git</li>
<li>twemoji-color-font</li>
<li>wallabag</li>
<li>wireless-regdb</li>
<li>yadm</li>
<li>zplug</li>
<li>zsh-bd</li>
<li>zsh-you-should-use</li>
</ul>
</details>

**Comment by siraben** [comment]
@SuperSandro2000 thanks, interestingly no log?

**Comment by SuperSandro2000** [comment]
> @SuperSandro2000 thanks, interestingly no log?

The blas failure is above and rustc somehow died in genereating man pages. I think I have built it before with 10 minute timeouts which worked. All other failures are somewhere intermediate and my logging is a bit broken so it could have failed fetching them.

**Comment by siraben** [comment]
@Ericson2314 If I don't see `stdenv.cc.cc` being referenced in an expression, and see `dontBuild = true`, is it safe to rewrite it to `stdenvNoCC.mkDerivation`? What other things should I check?

**Comment by Ericson2314** [comment]
@siraben I like the look of this :).

> What other things should I check?

Off the top of my head, `stdenv.cc` brings in:
 - `stdenv.cc.cc` as you say
 - `stdenv.cc.bintools`
 - `stdenv.cc.libc` the *binaries* that come with this like `ldd` can be used on their own.

**Comment by siraben** [comment]
Infinite recursion encountered

```
$ nix-env --query --available --out-path --file ./. --show-trace
error: infinite recursion encountered
```

**Comment by siraben** [comment]
Closing due to large number of merge conflicts.

**Pull request closed without merging**